rust/noosphere-sphere/src/authority/write.rs

```diff
@@ -0,0 +1,385 @@
+use crate::{internal::SphereContextInternal, HasMutableSphereContext, HasSphereContext};
+
+use super::SphereAuthorityRead;
+use anyhow::{anyhow, Result};
+use async_trait::async_trait;
+use cid::Cid;
+use noosphere_core::{
+    authority::{generate_capability, Authorization, SphereAction, SphereReference},
+    data::{DelegationIpld, Did, Jwt, Link, RevocationIpld},
+    view::SPHERE_LIFETIME,
+};
+use noosphere_storage::{Storage, UcanStore};
+use tokio_stream::StreamExt;
+use ucan::{
+    builder::UcanBuilder,
+    capability::{Capability, Resource, With},
+    crypto::KeyMaterial,
+};
+
+/// Any type which implements [SphereAuthorityWrite] is able to manipulate the
+/// [AuthorityIpld] section of a sphere. This includes authorizing other keys
+/// and revoking prior authorizations.
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+pub trait SphereAuthorityWrite<S>: SphereAuthorityRead<S>
+where
+    S: Storage + 'static,
+{
+    /// Authorize another key by its [Did], associating the authorization with a
+    /// provided display name
+    async fn authorize(&mut self, name: &str, identity: &Did) -> Result<Authorization>;
+
+    /// Revoke a previously granted authorization.
+    ///
+    /// Note that correctly revoking an authorization requires signing with a credential
+    /// that is in the chain of authority that ultimately granted the authorization being
+    /// revoked. Attempting to revoke a credential with any credential that isn't in that
+    /// chain of authority will fail.
+    async fn revoke_authorization(&mut self, authorization: &Authorization) -> Result<()>;
+
+    /// Recover authority by revoking all previously delegated authorizations
+    /// and creating a new one that delegates authority to the specified key
+    /// (given by its [Did]).
+    ///
+    /// Note that correctly recovering authority requires signing with the root
+    /// sphere credential, so generally can only be performed on a type that
+    /// implements [SphereAuthorityEscalate]
+    async fn recover_authority(&mut self, new_owner: &Did) -> Result<Authorization>;
+}
+
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+impl<C, S> SphereAuthorityWrite<S> for C
+where
+    C: HasSphereContext<S> + HasMutableSphereContext<S>,
+    S: Storage + 'static,
+{
+    // TODO(#423): We allow optional human-readable names for authorizations,
+    // but this will bear the consequence of leaking personal information about
+    // the user (e.g., a list of their authorized devices). We should encrypt
+    // these names so that they are only readable by the user themselves.
+    async fn authorize(&mut self, name: &str, identity: &Did) -> Result<Authorization> {
```

Comment:
"Authorize" is granting permission from some authority.
The `Authorization` struct represents some capability has been authorized.

Here, `authorize` means authorizing specifically the `SphereAction::Authorize` capability, which clouds the two usages for me. Separately, we check the author has some authorization and that is valid, but where is the check that the author specifically has the `Authorize` capability? It looks like the UCAN is created, authorizing some identity with `Authorize`, but do we check that, or rely on UCAN proof chains failing here? Or are these only available on "sudo spheres"?

Reply:
We rely on UCAN proof chains failing to pass muster in the future.